tests/kafkatest/tests/core/transactions_mixed_versions_test.py

```diff
@@ -0,0 +1,226 @@
+                   err_msg="Consumer failed to consume any messages for %ds" %\
+                   60)
+        return consumer
+
+    def drain_consumer(self, consumer, num_messages):
+        # wait until we read at least the expected number of messages.
+        # This is a safe check because both failure modes will be caught:
+        #  1. If we have 'num_seed_messages' but there are duplicates, then
+        #     this is checked for later.
+        #
+        #  2. If we never reach 'num_seed_messages', then this will cause the
+        #     test to fail.
+        wait_until(lambda: len(consumer.messages_consumed[1]) >= num_messages,
+                   timeout_sec=90,
+                   err_msg="Consumer consumed only %d out of %d messages in %ds" %\
+                   (len(consumer.messages_consumed[1]), num_messages, 90))
+        consumer.stop()
+        return consumer.messages_consumed[1]
+
+    def copy_messages_transactionally(self, input_topic, output_topic,
+                                      num_copiers, num_messages_to_copy,
+                                      use_group_metadata, group_protocol):
+        """Copies messages transactionally from the seeded input topic to the
+        output topic.
+
+        This method also consumes messages in read_committed mode from the
+        output topic.
+
+        It returns the concurrently consumed messages.
+        """
+
+        copiers = create_and_start_copiers(test_context=self.test_context,
+                                           kafka=self.kafka,
+                                           consumer_group=self.consumer_group,
+                                           input_topic=input_topic,
+                                           output_topic=output_topic,
+                                           transaction_size=self.transaction_size,
+                                           transaction_timeout=self.transaction_timeout,
+                                           num_copiers=num_copiers,
+                                           use_group_metadata=use_group_metadata)
+        concurrent_consumer = self.start_consumer(output_topic,
+                                                  group_id="concurrent_consumer",
+                                                  group_protocol=group_protocol)
+
+        copier_timeout_sec = 120
+        for copier in copiers:
+            wait_until(lambda: copier.is_done,
+                       timeout_sec=copier_timeout_sec,
+                       err_msg="%s - Failed to copy all messages in  %ds." %\
+                       (copier.transactional_id, copier_timeout_sec))
+        self.logger.info("finished copying messages")
+
+        return self.drain_consumer(concurrent_consumer, num_messages_to_copy)
+
+    def setup_topics(self):
+        assignment = ":".join(map(str, [self.kafka.idx(node) for node in self.kafka.nodes]))
+        transaction_assignment = ",".join(map(str, [assignment[::-1]] * 50))
```

Comment:
Let me check if I got this right. You want the leaders of input/output to be on kafka-0 -- the broker with the latest version -- and the transaction state partition to be on the last broker. Is it correct?

Reply:
Yes we want to send a verification request to a broker that might not support it.